tensorflow/python/kernel_tests/nn_ops/embedding_ops_test.py

```diff
@@ -909,6 +913,86 @@ def test_incompatible_types(self):
           sp_weights,
       )
 
+  def _SortByKey(self, keys, vals):
+    perm = sort_ops.argsort(keys)
+    return array_ops.gather(keys, perm), array_ops.gather(vals, perm)
+
+  def _ExpectedSparseGradient(
+          self, nnz, param_shape, np_type, sp_ids, sp_weights, combiner):
+    """Returns the expected indices and values corresponding to the (sparse)
+    gradient of a sparse embedding lookup.
+    """
+    expected_values = np.ones([nnz] + param_shape, dtype=np_type)
+    segment_ids = sp_ids.indices[:, 0]
+    ignore_weights = sp_weights is None
+    weights = (
+        array_ops.ones(nnz, dtype=dtypes.float32) if ignore_weights else
+        sp_weights.values)
+    if combiner == 'sqrtn':
+      weights = weights**2
+    segment_weights = math_ops.segment_sum(weights, segment_ids)
+    if combiner != 'sum':
+      grad_scale = 1.0 / array_ops.gather(segment_weights, segment_ids)
+      if combiner == 'sqrtn':
+        grad_scale = math_ops.sqrt(grad_scale)
+      expected_values *= grad_scale[:, None]
+    if not ignore_weights:
+      expected_values *= sp_weights.values[:, None]
+    expected_indices = sp_ids.values
+    # Sort and deduplicate the indices in the expected sparse tensor.
+    expected_indices, expected_values = self._SortByKey(
+            expected_indices, expected_values)
+    expected_indices, unique_mapping = array_ops.unique(expected_indices)
+    expected_values = math_ops.segment_sum(expected_values, unique_mapping)
+    return expected_indices, expected_values
+
+  def testResourceVariableGradientEmbeddingLookupSparse(self):
+    """Explicitly checks the gradient of a sparse embedding lookup with
+    ResourceVariable input."""
+    vocab_size = 128
+    batch_size = 32
+    param_shape = [16]
+    sp_ids, sp_weights, _, _, _ = (self._RandomIdsAndWeights(
+        batch_size, vocab_size))
+
+    for combiner, dtype, ignore_weights in itertools.product(
+            ["sum", "mean", "sqrtn"],
+            [dtypes.float32, dtypes.float64],
+            [True, False]):
+      with self.test_session(), \
+           forward_compat.forward_compatibility_horizon(2023, 9, 18):
+        x_shape = [vocab_size] + param_shape
+        np_type = "f" if dtype == dtypes.float32 else "d"
+        x = np.random.uniform(size=x_shape).astype(np_type) + 1
+        x = resource_variable_ops.ResourceVariable(x)
+
+        def forward(x_):
+          y_ = embedding_ops.embedding_lookup_sparse(
```

Comment:
I'm getting a failure on CPU tests around here:
```
bazel test //tensorflow/python/kernel_tests/nn_ops:embedding_ops_test_cpu --compilation_mode=opt --copt=-UNDEBUG
```
```
Detected at node 'embedding_lookup_sparse/embedding_lookup' defined at (most recent call last):
    ...
    File "tensorflow/python/kernel_tests/nn_ops/embedding_ops_test.py", line 978, in testResourceVariableGradientEmbeddingLookupSparse
    File "tensorflow/python/kernel_tests/nn_ops/embedding_ops_test.py", line 970, in forward

Node: 'embedding_lookup_sparse/embedding_lookup'
Container localhost does not exist. (Could not find resource: localhost/Variable)
	 [[{{node embedding_lookup_sparse/embedding_lookup}}]]
```

Reply:
I could only reproduce this by running under `@test_util.run_deprecated_v1`, but I've added the required variable initialization line to fix it.